Frameworks/ImageIO/CGImageDestination.mm

```diff
@@ -15,11 +16,867 @@
+                        toConsumer };
+
+@implementation ImageDestination
+
+- (instancetype)initToDestination:(size_t)frames
+                             type:(CFStringRef)type
+                             data:(CFMutableDataRef)data
+                              url:(CFURLRef)url {
+    if (self = [super init]) {
+        RETURN_NULL_IF(url && ![(NSURL*)url isFileURL]);
+        RETURN_NULL_IF(!data && !url);
+
+        self.maxCount = frames;
+
+        MULTI_QI imageQueryInterface = {0};
+        static const GUID IID_IWICImagingFactory = {0xec5ec8a9,0xc395,0x4314,0x9c,0x77,0x54,0xd7,0xa9,0x35,0xff,0x70};
+        imageQueryInterface.pIID = &IID_IWICImagingFactory;
+        RETURN_NULL_IF_FAILED(
+            CoCreateInstanceFromApp(CLSID_WICImagingFactory, nullptr, CLSCTX_INPROC_SERVER, nullptr, 1, &imageQueryInterface));
+        
+        _idFactory = (IWICImagingFactory*)imageQueryInterface.pItf;
+        RETURN_NULL_IF_FAILED(_idFactory->CreateStream(&_idStream));
+
+        if (CFStringCompare(type, kUTTypeJPEG, NULL) == kCFCompareEqualTo) {
+            _type = typeJPEG;
+            RETURN_NULL_IF_FAILED(_idFactory->CreateEncoder(GUID_ContainerFormatJpeg, NULL, &_idEncoder));
+        } else if (CFStringCompare(type, kUTTypeTIFF, NULL) == kCFCompareEqualTo) {
+            _type = typeTIFF;
+            RETURN_NULL_IF_FAILED(_idFactory->CreateEncoder(GUID_ContainerFormatTiff, NULL, &_idEncoder));
+        } else if (CFStringCompare(type, kUTTypeGIF, NULL) == kCFCompareEqualTo) {
+            _type = typeGIF;
+            RETURN_NULL_IF_FAILED(_idFactory->CreateEncoder(GUID_ContainerFormatGif, NULL, &_idEncoder));
+        } else if (CFStringCompare(type, kUTTypePNG, NULL) == kCFCompareEqualTo) {
+            _type = typePNG;
+            RETURN_NULL_IF_FAILED(_idFactory->CreateEncoder(GUID_ContainerFormatPng, NULL, &_idEncoder));
+        } else if (CFStringCompare(type, kUTTypeBMP, NULL) == kCFCompareEqualTo) {
+            _type = typeBMP;
+            RETURN_NULL_IF_FAILED(_idFactory->CreateEncoder(GUID_ContainerFormatBmp, NULL, &_idEncoder));
+        } else {
+            _type = typeUnknown;
+            return NULL;
+        }
+
+        if (url && [(NSURL*)url path]) { // Make sure that the url path actually resolves to something to avoid causing an exception
+            NSString* urlNSString = [[(NSURL*)url path] substringFromIndex:1];
+            NSData* urlAsData = [urlNSString dataUsingEncoding:NSUnicodeStringEncoding];
+            std::wstring wideUrl((wchar_t*)[urlAsData bytes], [urlAsData length]/sizeof(wchar_t));
+            RETURN_NULL_IF_FAILED(_idStream->InitializeFromFilename(wideUrl.c_str(), GENERIC_WRITE));
+
+            if (![(NSURL*)url checkResourceIsReachableAndReturnError:NULL]) {
+                return NULL;
+            }
+        } else if (data) {
+            _outData = data;
+
+            IStream* dataStream;
+            CreateStreamOnHGlobal(NULL, true, &dataStream);
```

Comment:
If you mean using an HRESULT or something, CreateStreamOnHGlobal does not have an HRESULT and I don't believe it can actually fail. If not, could you elaborate on what I should check for?